### PR TITLE
Refactor menu states and shortcuts in GMainWindow.

### LINKED ISSUE #5144: Slight Feature Removal. (Loading Amiibo's hotkey)
It does not bother me any but I've noticed the 'load amiibo' F2 hotkey that doesn't quite work is still listed under hotkeys. If it's a placeholder that's fine but it could get some user's excited about being able to load amiibo's to find it doesn't quite work as expected.

image

It's such a slight thing so feel free to close if it's not something worth approaching. I just figured it would save people asking if it was just outright removed for now. Thank you for your time.

### LINKED ISSUE #6043: Be able to exit emulator when using fullscreen (hotkeys don't work)
So that's it, a major feature is missing: be able tu use the hotkeys to control UI of the emu while in fullscreen; you cannot exit the emulation when playing in fullscreen if you first don't press ESC to go to window mode. This is annoying and is a basic feature for people like me using the emu with for example, Steam.

### LINKED ISSUE #7257: the F6 restart key doesn't work in yuzu mainline

### LINKED ISSUE #7429: F6 (Restart) will crash yuzu
BUG:

1. set View > Fullscreen + Single Window Mode
2. run a game
3. press F6 to restart

yuzu will close/crash (Italian Windows report):

```
Nome dell'applicazione che ha generato l'errore: yuzu.exe, versione: 0.0.0.0, timestamp: 0x619cf259
Nome del modulo che ha generato l'errore: yuzu.exe, versione: 0.0.0.0, timestamp: 0x619cf259
Codice eccezione: 0xc0000005
Offset errore 0x00000000002443be
ID processo che ha generato l'errore: 0x69a8
Ora di avvio dell'applicazione che ha generato l'errore: 0x01d7e121d37d4ee1
Percorso dell'applicazione che ha generato l'errore: C:\Users\Luca\Yuzu\yuzu.exe
Percorso del modulo che ha generato l'errore: C:\Users\Luca\Yuzu\yuzu.exe
```

I use fullscreen mode = exclusive in general config.
yuzu-windows-msvc-20211123-abd99b2af

_Gaming laptop ASUS GL703GS, Win 10, i7-8750H (6+6 cores), NVidia 1070 8GB, 32GB RAM, 1920x1080 144Hz G-Sync laptop screen, NVMe + SSD_

### PR DESCRIPTION
**Pull request by heinermann**
- Removed "Start", since it was always disabled unless it was "Continue" which has now been moved to "Pause".
- Fixes #6043
- Fixes #7429 
- Fixes #7257 
- Fixed #5144 

----

Also question: Does anyone have issue with renaming "Continue" to either "Unpause" or "Resume"?

### CODE REVIEW

File: src/yuzu/main.h [hunk 1]
```diff
@@ -9,6 +9,7 @@
 #include <unordered_map>
 
 #include <QMainWindow>
+#include <QShortcut>
```
Comment (Morph1984) [inline src/yuzu/main.h:12]: This can be replaced with a `QAction` forward declaration

Comment (heinermann) [reply src/yuzu/main.h]: Accidentally left this import in sorry, removed it.

File: src/yuzu/main.cpp [hunk 2]
```diff
@@ -952,164 +952,80 @@ void GMainWindow::InitializeRecentFileMenuActions() {
...
+    LinkActionShortcut(ui->action_Show_Filter_Bar, QStringLiteral("Toggle Filter Bar"));
+    LinkActionShortcut(ui->action_Show_Status_Bar, QStringLiteral("Toggle Status Bar"));
+    LinkActionShortcut(ui->action_Fullscreen, QStringLiteral("Fullscreen"));
+    LinkActionShortcut(ui->action_Capture_Screenshot, QStringLiteral("Capture Screenshot"));
+    LinkActionShortcut(ui->action_TAS_Start, QStringLiteral("TAS Start/Stop"));
+    LinkActionShortcut(ui->action_TAS_Record, QStringLiteral("TAS Record"));
+    LinkActionShortcut(ui->action_TAS_Reset, QStringLiteral("TAS Reset"));
+
+    static const QString main_window = QStringLiteral("Main Window");
+    const auto connect_shortcut = [&]<class Fn>(const QString& action_name, const Fn& function) {
```
Comment (Morph1984) [inline src/yuzu/main.cpp:981]:
```suggestion
    const auto connect_shortcut = [&]<typename Fn>(const QString& action_name, const Fn& function) {
```

File: src/yuzu/main.cpp [hunk 3]
```diff
@@ -1208,64 +1125,82 @@ void GMainWindow::ConnectWidgetEvents() {
 }
 
 void GMainWindow::ConnectMenuEvents() {
+    // Add actions to this window so that hiding menus in fullscreen won't disable them
+    const auto connect_to = [&]<class Fn>(QAction* action, const Fn& event_fn) {
```
Comment (Morph1984) [inline src/yuzu/main.cpp:1129]: ditto

File: src/yuzu/main.cpp [hunk 4]
```diff
@@ -1208,64 +1125,82 @@ void GMainWindow::ConnectWidgetEvents() {
...
-    connect(ui->action_Open_yuzu_Folder, &QAction::triggered, this, &GMainWindow::OnOpenYuzuFolder);
-    connect(ui->action_About, &QAction::triggered, this, &GMainWindow::OnAbout);
+    connect_to(ui->action_Open_yuzu_Folder, &GMainWindow::OnOpenYuzuFolder);
+    connect_to(ui->action_About, &GMainWindow::OnAbout);
+}
+
+void GMainWindow::UpdateMenuState() {
+    const bool is_paused = emu_thread == nullptr || !emu_thread->IsRunning();
+
+    const auto running_actions = {ui->action_Stop,
```
Comment (Morph1984) [inline src/yuzu/main.cpp:1186]:
```suggestion
    static constexpr std::array running_actions = {ui->action_Stop,
```

File: src/yuzu/main.cpp [hunk 5]
```diff
@@ -1208,64 +1125,82 @@ void GMainWindow::ConnectWidgetEvents() {
...
+                                  ui->action_Restart,
+                                  ui->action_Configure_Current_Game,
+                                  ui->action_Report_Compatibility,
+                                  ui->action_Load_Amiibo,
+                                  ui->action_Pause};
```
Comment (Morph1984) [inline src/yuzu/main.cpp:1191]: and run clang-format
```suggestion
                                  ui->action_Pause,};
```

Comment (heinermann) [reply src/yuzu/main.cpp:1191]: clang-format doesn't do anything here.

File: src/yuzu/main.cpp (hunk 4 above)
Comment (heinermann) [reply src/yuzu/main.cpp:1186]: `ui` is a member of `GMainWindow` so this won't work.

Comment (Morph1984) [reply src/yuzu/main.cpp:1186]: revised the review

Comment (heinermann) [reply src/yuzu/main.cpp:1186]: Is there a reason this is better?

Comment (Morph1984) [reply src/yuzu/main.cpp:1186]: auto will deduce to `std::initializer_list`, and we want to be explicit that this is an array of elements

Comment (heinermann) [reply src/yuzu/main.cpp:1186]: Sure, I just don't see that as a downside in this code specifically. The same mechanism is used elsewhere for iteration over a `std::initializer_list`.

File: src/yuzu/main.cpp (hunk 5 above)
Comment (heinermann) [reply src/yuzu/main.cpp]: Given its usage elsewhere I'll add the trailing comma but this isn't part of clang-format. Is that style just not included?

Comment (german77) [reply src/yuzu/main.cpp]: The whole yuzu codebase use a coma after the last element of arrays or enums. This allows for better clang formatting.
image

Comment (Morph1984) [reply src/yuzu/main.cpp]: Basically what narr said, I wouldn't have suggested this if it wasn't used everywhere for better formatting

Comment (heinermann) [reply src/yuzu/main.cpp]: Yep realized after I wrote what I wrote. This should be part of some kind of checkstyle if possible.

**Comment by german77** [comment]
This PR seems to break reloading the game(F6). With the following error. Master is working without issues.
```
[  39.068340] Kernel <Warning> core/hle/kernel/k_process.cpp:DecrementThreadCount:169: Process termination is not fully implemented.
[  39.068920] Kernel <Warning> core/hle/kernel/kernel.cpp:Shutdown:187: 226 kernel objects were dangling on shutdown!
[  39.076065] Common.Filesystem <Error> common/fs/path_util.cpp:ValidatePath:138: Input path is empty, path=
[  39.076070] Common.Filesystem <Error> common/fs/fs.cpp:FileOpen:134: Input path is not valid, path=
```

**Comment by german77** [comment]
I was able to reproduce. This error happens because `game_path` it's empty on `BootGame(game_path)`

**Comment by heinermann** [comment]
> This PR seems to break reloading the game(F6). With the following error. Master is working without issues.
> 
> ```
> [  39.068340] Kernel <Warning> core/hle/kernel/k_process.cpp:DecrementThreadCount:169: Process termination is not fully implemented.
> [  39.068920] Kernel <Warning> core/hle/kernel/kernel.cpp:Shutdown:187: 226 kernel objects were dangling on shutdown!
> [  39.076065] Common.Filesystem <Error> common/fs/path_util.cpp:ValidatePath:138: Input path is empty, path=
> [  39.076070] Common.Filesystem <Error> common/fs/fs.cpp:FileOpen:134: Input path is not valid, path=
> ```

Nope the same crash happens in master. Switch to fullscreen since the hotkey doesn't work in windowed then press F6. I'll fix it in a bit.

**Comment by german77** [comment]
It gets cleared on `ShutdownGame()`  with `game_path.clear();` In any case latest commit fixes the issue by creating a copy of the string before clearing the path

**Comment by heinermann** [comment]
I'm going through the issue tracker and linking additional issues.

**Review by german77 — APPROVED** [review]

**Comment by german77** [comment]
#5144 is not solved by your PR. Amiibos are broken in the emulation side not hotkey

**Comment by heinermann** [comment]
> #5144 is not solved by your PR. Amiibos are broken in the emulation side not hotkey

The issue raised was specifically for the hotkey, then the discussion changed at the bottom.

**Comment by heinermann** [comment]
Also need to fix hotkeys for non-single window mode, give me a moment.

**Review by Morph1984 — APPROVED** [review]

**Comment by german77** [comment]
I wanted to remove non-single window mode since it's so buggy. But they told me it was a nice feature, I still don't know who uses it. Good luck if you try to fix it.

**Pull request merged**